doc/dev/dialects.rst

```diff
@@ -0,0 +1,417 @@
+
+- a *C++* part: Additional hand-written logic, and most transformations, are defined directly in C++
+  source files. An example might be a storage class for a data type, or the implementation of
+  interface methods for an operation declared in ODS.
+  Note that while the usage of ODS is strongly recommended, full control over the underlying
+  objects can always be achieved by directly working with the relevant C++ classes.
+
+In Catalyst, hand-written implementations for classes & methods can be found in the source
+directory, under `catalyst/mlir/lib\/<DialectName\>/IR <https://github.com/PennyLaneAI/catalyst/tree/main/mlir/lib/Quantum/IR>`_.
+
+An alternative tutorial on creating a basic MLIR dialect can also be found in the official docs, see
+`Creating a Dialect <https://mlir.llvm.org/docs/Tutorials/CreatingADialect/>`_.
+
+
+Creating a new dialect
+======================
+
+Let's see how we can add a new dialect to Catalyst, using OpenQASM as an example! Note that all
+auto-generated files and build artifacts mentioned in the guide will only appear at the end of
+the `Using the dialect <#using-the-dialect>`_ section when the ``make dialects`` command is run.
+
+Start by creating a new TableGen file, located at ``mlir/include/OpenQASM/OpenQASM.td``, with the
+following content:
+
+.. code-block:: tablegen
+
+    include "mlir/IR/DialectBase.td"
+
+    def OpenQASMDialect : Dialect {
+        let summary = "An OpenQASM 3 dialect in MLIR.";
+        let description = [{
+            ...
+        }];
+
+        /// This is the namespace of the dialect in MLIR, which is used as a prefix for types and ops.
+        let name = "oq";
+
+        /// This is the C++ namespace that the dialect, and all sub-components, get placed in.
+        let cppNamespace = "::catalyst::openqasm";
+
+        /// Use the default type printing/parsing hooks, otherwise we would explicitly define them.
+        let useDefaultTypePrinterParser = 1;
+    }
+
+Note the use of ``def`` in TableGen indicates that we are defining a new type of object - under the
+hood this results in a new C++ class. To avoid repeating common parts of object definitions, this
+object creation mechanism itself supports classes. ``def OpenQASMDialect : Dialect`` defines an
+instance of the (built-in) ``Dialect`` TableGen type. In this way, ``OpenQASMDialect`` will inherit
+all properties defined for the ``Dialect`` type.
+
+We could also have subtyped the ``Dialect`` type with ``class QuantumDialect : Dialect``, and then
+created two kinds of quantum dialects as ``def OpenQASMDialect : QuantumDialect`` and
+``def QIRDialect : QuantumDialect`` - if we thought that the two shared some common elements.
+Objects defined with ``def`` are terminal and can no longer be subtyped or inherited from.
+
+Similarly, a new data type for our dialect can be added by via the built-in ``TypeDef`` class:
+
+.. code-block:: tablegen
+
+    include "mlir/IR/AttrTypeBase.td"
```

Comment:
are these three tablegen codeblocks different files?

Reply:
Same file, I just think it would be easier to explain the parts separately.

Reply:
ah okay, might be worth adding a comment on that (unless I missed it). The `include` made me think it was a new file (since you usually see them at the top of files)

Reply:
https://github.com/PennyLaneAI/catalyst/pull/943/commits/fbed8e4f8353c9465d63fb820948f9378cb52588